### ISSUE TITLE
Logging cruft

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#979]
Several of the logging views appear to accumulate cruft as the system runs. In some cases (`logs_histogram`) this is "working as intended" and we may need to change the intent; in other cases (`logs_frontiers`) there appear to be frontiers for dataflows that have long since departed, and are not correctly being cleaned up.

These can reach up into the millions of record, at least after a day of chbench exercise, and removing them would surely remove some memory growth and compute overhead.

**Comment by cuongdo** [comment]
@frankmcsherry are you actively looking at this?

**Comment by frankmcsherry** [comment]
The two logging sources that appear to hog memory are the `mz_scheduling_*` sources. They are a bit awkward because their contents are derived data, rather than source data, so we cannot as easily just retain the data on the side and discard it when dataflows close out (the strategy used for other logging sources we want to clean). 

What might work is the same technique as for monotonic sources, which is to have a feedback variable which we use to delete records that are no longer relevant. In particular, we can have as input both scheduling event data (e.g. elapsed times, histograms) as well as a stream of currently active operators. If we join those two to produce the "legitimate output" and feed back the actual input minus the legit output, and retract it from the input, we should clean up the measurements.

This relies on logging being sane with respect to timestamps, which I think it is (operator start up should happen before the first scheduling events, and shut down should happen after the last one).

The monotonic fix with feedback hasn't been flawless yet. There are still some resource consumption issues to shake out there (it does return to zero, but spikes up high for quite a while). So, while this proposal *should* work there may be some gotchas along the way.

**Comment by ruchirK** [comment]
@frankmcsherry - @cuongdo asked me to take this on. I think I understand the proposed approach here and can poke at it for a day or two, and after that maybe come back to you with more specific questions.

Does that sound good?

**Comment by frankmcsherry** [comment]
Works for me. I'm happy to help out with prototypes, or talking you through things!

**Comment by ruchirK** [comment]
I think we can close this now that #4127 has merged and we compact the heaviest logs away, and I can open more targeted issues for followups

**Comment by frankmcsherry** [comment]
I think that is right. We can certainly clean things up a bit more, but the main issue should be resolved (unless we hear that it is not).
